plugins/hls-explicit-imports-plugin/src/Ide/Plugin/ExplicitImports.hs

```diff
@@ -117,65 +136,99 @@ lensProvider
   CodeLensParams {_textDocument = TextDocumentIdentifier {_uri}}
     -- VSCode uses URIs instead of file paths
     -- haskell-lsp provides conversion functions
-    | Just nfp <- uriToNormalizedFilePath $ toNormalizedUri _uri = liftIO $
+    | Just nfp <- uriToNormalizedFilePath $ toNormalizedUri _uri =
       do
-        mbMinImports <- runAction "MinimalImports" state $ useWithStale MinimalImports nfp
+        mbMinImports <- liftIO $ runAction "MinimalImports" state $ useWithStale MinimalImports nfp
         case mbMinImports of
-          -- Implement the provider logic:
-          -- for every import, if it's lacking a explicit list, generate a code lens
-          Just (MinimalImportsResult minImports, posMapping) -> do
-            commands <-
-              sequence
-                [ generateLens pId _uri edit
-                  | (imp, Just minImport) <- minImports,
-                    Just edit <- [mkExplicitEdit pred posMapping imp minImport]
-                ]
-            return $ Right $ InL $ catMaybes commands
+          Just (MinimalImportsResult minImports _ _, posMapping) -> do
+            let lens = [ generateLens _uri curRange int
+                       | (range, int) <- minImports
+                       , let Just curRange = toCurrentRange posMapping range]
+            return $ Right $ InL lens
           _ ->
             return $ Right $ InL []
     | otherwise =
       return $ Right $ InL []
+  where generateLens :: Uri  -> Range -> Int -> CodeLens
+        generateLens uri range int =
+          CodeLens { _data_ = Just $ toJSON $ ResolveOne uri int
+                   , _range = range
+                   , _command = Nothing }
+
+lensResolveProvider :: PluginMethodHandler IdeState Method_CodeLensResolve
+lensResolveProvider ideState plId cl@(CodeLens {_data_ = Just data_}) = pluginResponse $ do
+  case fromJSON data_ of
+    Success (ResolveOne uri int) -> do
+      nfp <- getNormalizedFilePath uri
+      (MinimalImportsResult _ _ resolveMinImp) <-
+        handleMaybeM "Unable to run Minimal Imports"
+        $ liftIO
+        $ runAction "MinimalImports" ideState $ use MinimalImports nfp
+      target <- handleMaybe "Unable to resolve lens" $ resolveMinImp IM.!? int
+      let updatedCodeLens = cl & L.command ?~  mkCommand plId uri target data_
+      return updatedCodeLens
+    _ -> throwE "unable to parse data_ field of CodeLens"
```

Comment:
Error handling things: this doesn't tell us what it _was_, and also discards the error from aeson that might be helpful. Hopefully we will already have logged the CodeAction itself, so we'll be able to see the data field there, but it's usually helpful to repeat things. And we definitely want to reproduce the aeson error.

This would be easier with structured errors, but even just printing them is an improvement over getting no info!

(This applies in some of your other code too, I think)